src/node_native_module.cc

```diff
@@ -19,100 +19,115 @@ using v8::Isolate;
-  // std::map<std::string, const uint8_t*> instead of a v8::Object,
-  // because here they are only looked up from the C++ side
-  // (except in process.binding('natives') which we don't use)
-  // so there is little value to put them in a v8::Object upfront.
-  // Moreover, a std::map lookup should be faster than a lookup on
-  // an V8 Object in dictionary mode.
+void NativeModuleLoader::GetCacheUsage(
+    Local<Name> property, const PropertyCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
   Isolate* isolate = env->isolate();
   Local<Context> context = env->context();
-  Local<Value> null = Null(isolate);
-
-  Local<Object> native_modules_source = Object::New(isolate);
-  CHECK(native_modules_source->SetPrototype(context, null).FromJust());
-  DefineJavaScript(env, native_modules_source);
-  native_modules_source->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+  Local<Object> result = Object::New(isolate);
+  result
+      ->Set(env->context(),
+            OneByteString(isolate, "compiledWithCache"),
+            ToJsSet(context, env->native_modules_with_cache))
       .FromJust();
-  env->set_native_modules_source(native_modules_source);
-
-  Local<Object> native_modules_source_hash = Object::New(isolate);
-  CHECK(native_modules_source_hash->SetPrototype(context, null).FromJust());
-  DefineJavaScriptHash(env, native_modules_source_hash);
-  native_modules_source_hash
-      ->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
+  result
+      ->Set(env->context(),
+            OneByteString(isolate, "compiledWithoutCache"),
+            ToJsSet(context, env->native_modules_without_cache))
       .FromJust();
-  env->set_native_modules_source_hash(native_modules_source_hash);
+  info.GetReturnValue().Set(result);
+}
 
-  Local<Object> native_modules_code_cache = Object::New(isolate);
-  CHECK(native_modules_code_cache->SetPrototype(context, null).FromJust());
-  DefineCodeCache(env, native_modules_code_cache);
-  native_modules_code_cache->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
-      .FromJust();
-  env->set_native_modules_code_cache(native_modules_code_cache);
+void NativeModuleLoader::GetSourceObject(
+    Local<Name> property, const PropertyCallbackInfo<Value>& info) {
+  Environment* env = Environment::GetCurrent(info);
+  info.GetReturnValue().Set(per_process_loader.GetSourceObject(env->context()));
+}
 
-  Local<Object> native_modules_code_cache_hash = Object::New(isolate);
-  CHECK(native_modules_code_cache_hash->SetPrototype(context, null).FromJust());
-  DefineCodeCacheHash(env, native_modules_code_cache_hash);
-  native_modules_code_cache_hash
-      ->SetIntegrityLevel(context, IntegrityLevel::kFrozen)
-      .FromJust();
-  env->set_native_modules_code_cache_hash(native_modules_code_cache_hash);
+Local<Object> NativeModuleLoader::GetSourceObject(
+    Local<Context> context) const {
+  return MapToObject(context, source_);
```

Comment:
Is the return value cached / should it be cached? Right now it seems like every time `internalBinding('native_module').source` is accessed a new object is created.

Reply:
It's not cached, but that does not really matter since this is only used by `tools/generate_code_cache.js`